### DISCUSSION TITLE
Scripts with dynamic parameters

### DISCUSSION THREAD

**Discussion by dainok** [netbox-community/netbox#11531] [Q&A]
For the project I'm working on, I need to enqueue custom scripts with dynamic parameters.
Basically, after a user input, I need to launch time expensive scripts via RQ, capturing results, errors and so on. Those scripts requires some parameters, sent by the user.

I checked the Script doc (https://docs.netbox.dev/en/stable/customization/custom-scripts/) but seems I cannot follow that way.

I tried to reuse the script classes, to implement a third batch type (let's call it "task"), but I failed because _get_script is looking for a file.

I wonder if there is a way to do that, natively in NetBox.

Could you point me in the right direction?

**Comment by mtinberg** [comment]
What you seem to be describing is Scripts, why did you think they were not going to meet your needs?


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Andrea Dainese ***@***.***>
Sent: Wednesday, January 18, 2023 10:49 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Scripts with dynamic parameters (Discussion #11531)


For the project I'm working on, I need to enqueue custom scripts with dynamic parameters.
Basically, after a user input, I need to launch time expensive scripts via RQ, capturing results, errors and so on. Those scripts requires some parameters, sent by the user.

I checked the Script doc (https://docs.netbox.dev/en/stable/customization/custom-scripts/) but seems I cannot follow that way.

I tried to reuse the script classes, to implement a third batch type (let's call it "task"), but I failed because _get_script is looking for a file.

I wonder if there is a way to do that, natively in NetBox.

Could you point me in the right direction?

**Comment by dainok** [comment]
Hi @mtinberg !

Thank you for your answer.

After your suggestion I made few tests, I realize Script could fit my needs. I can now start a Script but I cannot see the log in the "Job Results" page. Probably I'm missing something. I'm programmatically starting the script from a View as following:

~~~
from myapp.scripts import MyJob

j = MyJob.MyJob()
j.run({'mydata': 'data'}, True)
~~~

Another question if you don't mind: any chance to put my scripts inside my plugin and having them recognized by NetBox without moving them in the `netbox/scripts` folder?

**Reply by candlerb** [reply]
Scripts don't need to start jobs - they run as a job themselves.

If you're writing a plugin, then you should ignore scripts, make your own data entry form, and run your own background tasks as described [here](https://docs.netbox.dev/en/stable/plugins/development/background-tasks/). 

In other words: you should write your code *either* as a standalone custom script, *or* as a plugin.  A script may be sufficient if your requirements are very simple: just present a data form, and then run some code triggered by that form.  A plugin is for more complex requirements.

**Reply by dainok** [reply]
Thank you for your advice. I'll review my approach.
